lib/pure/strutils.nim

```diff
@@ -2050,97 +2050,152 @@ func contains*(s: string, chars: set[char]): bool =
   ## * `find func<#find,string,set[char],Natural,int>`_
   return find(s, chars) >= 0
 
-func replace*(s, sub: string, by = ""): string {.rtl,
+func replace*(s: string, sub, by: char, max = -1): string {.rtl,
+    extern: "nsuReplaceChar".} =
+  ## Replaces `max` occurrences of the character `sub` in `s` with
+  ## the character `by`.
+  ## If `max == -1`, it replaces all occurrences.
+  ##
+  ## Optimized version of `replace <#replace,string,string,string>`_ for
+  ## characters.
+  ##
+  ## See also:
+  ## * `find func<#find,string,char,Natural,int>`_
+  ## * `replaceWord func<#replaceWord,string,string,string>`_
+  ## * `multiReplace func<#multiReplace,string,varargs[]>`_
+  runnableExamples:
+    assert "valid variable name".replace(' ', '_') == "valid_variable_name"
+    assert "Faabar!".replace('a', 'o', 2) == "Foobar!"
+  result = newStringOfCap(s.len)
+  var
+    i = 0
+    occLeft = max
+  while i <= s.high:
+    if occLeft == 0: break
+    if s[i] == sub:
+      result.add by
+      if occLeft > 0: dec occLeft
+    else:
+      result.add s[i]
+    inc i
+  if occLeft == 0:
+    # copy the rest:
+    result.add s.substr(i)
+
+# Maybe implement something like this for all replace functions...?
```

Comment:
@kintrix007 Why do we have this commented block here? Should this instead be a separate issue or RFC?

Reply:
We do not.
I left it in case someone had something to say about it.
Will definitely remove it if the PR gets accepted.
But now that you mention it, I should have probably just said that as a comment...

Reply:
> Will definitely remove it if the PR gets accepted.

If the PR gets accepted, someone will just merge it. So it's better to keep the PR in the state you want to see it merged.

Reply:
also, the implementation is not good (no early break), so if you want to keep a comment just add:
```nim
# consider adding: func replace*(s: string, subs: set[char], by: char, max = -1): string 
```
(whether it's a good idea or not, is unclear; the best algorithms to write in stdlib are the ones that are tricky to implement correctly/efficiently in user code, ones that simplify user code in common cases, etc)